### PR TITLE
WIP: Github Enterprise URL Addition

### PR DESCRIPTION
**Pull request by kodaman2**
This is an enhancement as per #75, it adds an input text box for GHE Url that goes to the store settings. If there is a URL in this text box, then octokit uses the custom URL given for authentication.

@flashadvocate please download/clone this commit https://github.com/kodaman2/code-notes/tree/ghe-token:

- npm install
- npm run dev

I am having hard time finding documentation on the correct url for ghe please try the below, I assume it should be secured http requests:

- http://github.ncsu.edu
- https://github.ncsu.edu
- http://github.ncsu.edu/api/v3
- https://github.ncsu.edu/api/v3

Then add your personal token, and let me know how it goes, and which if any url worked.

> Side note: I had to downgrade webpack-dev-server as per #92 in order to `npm run dev`

### CODE REVIEW

**Comment by lauthieb** [comment]
Hi! FYI, the commit from @jack-chapman is now merged on `master` branch :)

**Comment by kodaman2** [comment]
> Hi! FYI, the commit from @jack-chapman is now merged on `master` branch :)

Latest master has been merged, and still awaiting a courageous enterprise user to test the PR lol.

**Comment by lauthieb** [comment]
I don't use Github Enterprise sorry :(

**Comment by flashadvocate** [comment]
Sorry sorry, I'll give it a go. Moving fast I see :)

**Comment by flashadvocate** [comment]
@kodaman2 what scopes does it need? (For the token)

**Comment by flashadvocate** [comment]
@kodaman2 getting `bad credentials`

```
Uncaught (in promise) HttpError: {"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}
    at IncomingMessage.response.on (/Users/dcdeaton/Projects/code-notes/node_modules/@octokit/rest/lib/request/request.js:73:18)
    at emitNone (events.js:111:20)
    at IncomingMessage.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1056:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

I am able to authenticate to the GHE server with my token though
`curl -u username:<PAT> https://github.ncsu.edu/api/v3/gists`
```
[
  {
    "url": "https://github.ncsu.edu/api/v3/gists/5ddd3d58a5e681de69118ddee425e05e",
    "forks_url": "https://github.ncsu.edu/api/v3/gists/5ddd3d58a5e681de69118ddee425e05e/forks",
    "commits_url": "https://github.ncsu.edu/api/v3/gists/5ddd3d58a5e681de69118ddee425e05e/commits",
    "id": "5ddd3d58a5e681de69118ddee425e05e",
...
```

**Comment by flashadvocate** [comment]
Not sure what's going on here... thought i might have found the issue, where the username was missing in the octokit authentication call, but so far no dice.

I've tried...

```
octokit.authenticate({
  type: 'basic',
  username: 'my@email.com',
  password: '<my personal access token>'
});
```

```
octokit.authenticate({
  type: 'basic',
  username: 'myusername',
  password: '<my personal access token>'
});
```

```
octokit.authenticate({
  type: 'token',
  username: 'username',
  password: '<my personal access token>'
});
```

Nada

**Comment by kodaman2** [comment]
Thanks for the update. I think the best thing to do here is to create a small node project and test ghe authentication first, working on that today, I'll post repo link later on.

The last authenticate example on your last comment did you tried exactly like that? Can you try changing password to token, and hardcode your username:

```
octokit.authenticate({
  type: 'token',
  username: 'username',
  token: '<my personal access token>'
});
```

**Comment by kodaman2** [comment]
Ok I think I got to the bottom of this, hopefully lol. As is looks octokit/rest documentation site is only for version 16+ and you can't filter for older versions hence why this is not working, there has been tons of changes some which include parameter names and function names. I downloaded octokit/rest 14.0.8 the version code-notes is using and looked at the source code, and got a bit lucky that version had an example for github enterprise, thought I am not sure if only oauth auth works or token so I wrote two js scripts to see which one works or if both do.

https://github.com/kodaman2/octokit-api-test

> Paste your token in the script before running it.
```
npm install
node index.js // this will test token auth
node enterprise.js // this will test oauth
```
If that doesn't yield results a good idea might be to do vscode live share, and do a little remote debugging.

**Comment by kodaman2** [comment]
As a side note this is how much the interface changed from 14 to 16, that's why baseUrl wasn't working  😃, if latest.js works, I'll submit another PR to upgrade octokit version:

v14.0.8:
```js
export interface Options {
    timeout?: number;
    host?: string;
    pathPrefix?: string;
    protocol?: string;
    port?: number;
    proxy?: string;
    ca?: string;
    headers?: {[header: string]: any};
    requestMedia?: string;
    rejectUnauthorized?: boolean;
    family?: number;
  }
```

v16.32.0:
```js
export interface Options {
    auth?:
      | string
      | { username: string; password: string; on2fa: () => Promise<string> }
      | { clientId: string; clientSecret: string }
      | { (): string | Promise<string> };
    userAgent?: string;
    previews?: string[];
    baseUrl?: string;
    log?: {
      debug?: (message: string, info?: object) => void;
      info?: (message: string, info?: object) => void;
      warn?: (message: string, info?: object) => void;
      error?: (message: string, info?: object) => void;
    };
    request?: {
      agent?: http.Agent;
      timeout?: number;
    };
    timeout?: number; // Deprecated
    headers?: { [header: string]: any }; // Deprecated
    agent?: http.Agent; // Deprecated
    [option: string]: any;
  }
```

**Comment by flashadvocate** [comment]
@kodaman2 latest worked. The other two did not

**Comment by kodaman2** [comment]
> @kodaman2 latest worked. The other two did not

ok, awesome, I'll work on upgrading octokit/rest then, and re-editing the request on this repo for ghe. That's good progress!

**Comment by kodaman2** [comment]
@flashadvocate could you retry `index.js` and `enterprise.js`, I've updated the url without the prefix parameter.

revert octokit/rest for those two.
```
npm i @octokit/rest@14.0.8
```

**Comment by flashadvocate** [comment]
Results running both `index.js` and `enterprise.js`

```
[before-after-hook]: "Hook()" repurposing warning, use "Hook.Collection()". Read more: https://git.io/upgrade-before-after-hook-to-1.4
(node:20898) UnhandledPromiseRejectionWarning: Error: getaddrinfo ENOTFOUND https https:443
    at ClientRequest.request.on (/Users/dcdeaton/Projects/octokit-api-test/node_modules/@octokit/rest/lib/request/request.js:93:14)
    at ClientRequest.emit (events.js:198:13)
    at TLSSocket.socketErrorListener (_http_client.js:392:9)
    at TLSSocket.emit (events.js:198:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

**Comment by kodaman2** [comment]
ok, well nvm I'll go ahead and update the package, is probably for the better anyways. Thanks for the results, I'll let you know when I integrate it on the app.

**Comment by flashadvocate** [comment]
No problem, thanks for the work. VueJS and electron seem to be a nice combination :)

**Comment by kodaman2** [comment]
No worries, I actually use gists quite a lot so I thought it was just to contribute to this repo.

**Comment by kodaman2** [comment]
Latest commit:

- upgraded octokit/rest from 14.0.8 to 16.32.0 as 14.0.8 was not working at all with ghe auth
- minor changes on octokit.gists calls, most functions stayed the same except edit was changed to update, and parameter id was changed to gists_id
- ran lint checks

@flashadvocate tell me something good : ) I am hoping we are close to completion on this PR
Please fetch/clone latest commit on this branch:

```
npm install // ensure octokit/rest is upgraded to 16.32.0 on package.json
npm run dev
```

**Comment by flashadvocate** [comment]
@kodaman2 congrats. You've got a working implementation there ;)

**Comment by kodaman2** [comment]
@flashadvocate awesome! thanks for the update.

@lauthieb I believe this is ready for review, and merge. If you want to double check and make sure I didn't miss any octokit.gists calls but I believe I covered all my basis. : )

File: src/renderer/components/modals/settings-modal/SettingsModal.html [hunk 1]
```diff
@@ -5,6 +5,14 @@
...
 
     <section class="modal-card-body">
+        <b-field label="Github Enterprise Users Only">
+            <b-input
+              type="text"
+              ref="githubEnterpriseUrl"
+              v-model="newSettings.githubEnterpriseUrl"
+              placeholder="https://github.host/api/v3">
+            </b-input>
+      </b-field>
```
Comment (lauthieb) [inline src/renderer/components/modals/settings-modal/SettingsModal.html]: Please, fix indentation :)

File: src/renderer/store/modules/Note.js [hunk 2]
```diff
@@ -4,12 +4,8 @@ import path from 'path';
...
 import { remote } from 'electron';
 
-const octokit = require('@octokit/rest')({
-  requestMedia: 'application/vnd.github.v3+json',
-  headers: {
-    'user-agent': 'octokit/rest.js v1.2.3',
-  },
-});
+const Octokit = require('@octokit/rest');
+let octokit = Octokit();
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: Are you sure?
In their documentation I see:
```
const Octokit = require("@octokit/rest");
const octokit = new Octokit();
```

Comment (lauthieb) [reply src/renderer/store/modules/Note.js]: You missed the `new` keyword.

File: src/renderer/store/modules/Note.js [hunk 3]
```diff
@@ -53,16 +49,30 @@ const actions = {
...
         store.commit('LOAD_NOTES', []);
 
-        octokit.authenticate({
-          type: 'token',
-          token: store.rootState.Settings.settings.githubPersonalAccessToken,
-        });
+        if (store.rootState.Settings.settings.githubEnterpriseUrl) {
+          octokit = Octokit({
+            baseUrl: store.rootState.Settings.settings.githubEnterpriseUrl,
+            userAgent: 'code-notes v1.2.3',
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: Do you think that we can get the "v1.2.3" version from our `package.json` file?

File: src/renderer/store/modules/Note.js [hunk 4]
```diff
@@ -53,16 +49,30 @@ const actions = {
...
+            auth: store.rootState.Settings.settings.githubPersonalAccessToken,
+            mediaType: {
+              format: 'application/vnd.github.v3+json',
+            },
+          });
+        } else {
+          octokit = Octokit({
+            userAgent: 'code-notes v1.2.3',
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: same here

File: src/renderer/store/modules/Note.js [hunk 5]
```diff
@@ -53,16 +49,30 @@ const actions = {
...
+            auth: store.rootState.Settings.settings.githubPersonalAccessToken,
+            mediaType: {
+              format: 'application/vnd.github.v3+json',
+            },
+          });
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: This code could be simplified if you set the octokit object before and then just add a `baseUrl` property only if `store.rootState.Settings.settings.githubEnterpriseUrl` is present.

File: src/renderer/components/modals/settings-modal/SettingsModal.html (hunk 1 above)
Comment (kodaman2) [reply src/renderer/components/modals/settings-modal/SettingsModal.html]: Yes, will do.

File: src/renderer/store/modules/Note.js (hunk 2 above)
Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: I agree.

File: src/renderer/store/modules/Note.js (hunk 3 above)
Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: Yes, that is a valid point. I thought the user agent was something related to octokit/rest internal api, but upon reading github api docs, yes this should be code-notes + version, will fix it in both places.

File: src/renderer/store/modules/Note.js (hunk 4 above)
Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: Answer same as above.

File: src/renderer/store/modules/Note.js (hunk 5 above)
Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: Yes I agree, this is due to my noob skills in java (python/java dev here), will def clean up as per your suggestion.

**Comment by kodaman2** [comment]
@lauthieb code changes requested have been fixed, was able to resolved all. Let me know what you think on the Octokit constructor fix, and I am open to suggestions I haven't been doing javascript for too long.

@flashadvocate functionality should work the same but please test to make sure, 8bf62f864dac622a493de9c7f3ec89db34156daa

File: src/renderer/store/modules/Note.js [hunk 6]
```diff
@@ -3,11 +3,13 @@ import converter from '@/converter';
...
 // eslint-disable-next-line import/no-extraneous-dependencies
 import { remote } from 'electron';
+import packageJson from '../../../../package';
 
-const octokit = require('@octokit/rest')({
-  requestMedia: 'application/vnd.github.v3+json',
-  headers: {
-    'user-agent': 'octokit/rest.js v1.2.3',
+const Octokit = require('@octokit/rest');
+let octokit = new Octokit({
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: I think you can remove the `new Ocktokit` and do this into the if/else statement below.

File: src/renderer/store/modules/Note.js [hunk 7]
```diff
@@ -53,16 +55,20 @@ const actions = {
         store.commit('SELECT_LOADING', true);
         store.commit('LOAD_NOTES', []);
 
-        octokit.authenticate({
-          type: 'token',
-          token: store.rootState.Settings.settings.githubPersonalAccessToken,
+        if (store.rootState.Settings.settings.githubEnterpriseUrl) {
+          octokit = Octokit({
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: `Octokit` => `new Octokit`

File: src/renderer/store/modules/Note.js [hunk 8]
```diff
@@ -53,16 +55,20 @@ const actions = {
...
+            baseUrl: store.rootState.Settings.settings.githubEnterpriseUrl,
+          });
+        }
+        octokit = Octokit({
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: `Octokit` => `new Octokit`

File: src/renderer/store/modules/Note.js [hunk 9]
```diff
@@ -53,16 +55,20 @@ const actions = {
         store.commit('SELECT_LOADING', true);
         store.commit('LOAD_NOTES', []);
 
-        octokit.authenticate({
-          type: 'token',
-          token: store.rootState.Settings.settings.githubPersonalAccessToken,
+        if (store.rootState.Settings.settings.githubEnterpriseUrl) {
+          octokit = Octokit({
+            baseUrl: store.rootState.Settings.settings.githubEnterpriseUrl,
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: You missed the `auth` property here ;)

File: src/renderer/store/modules/Note.js [hunk 10]
```diff
@@ -3,11 +3,13 @@ import converter from '@/converter';
 import path from 'path';
 // eslint-disable-next-line import/no-extraneous-dependencies
 import { remote } from 'electron';
+import packageJson from '../../../../package';
```
Comment (lauthieb) [inline src/renderer/store/modules/Note.js]: Nice :D

File: src/renderer/store/modules/Note.js (hunk 6 above)
Comment (lauthieb) [reply src/renderer/store/modules/Note.js]: Just declare a simple object

Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: Not sure I follow. So, like this?:

```
let octokit = Octokit({
  userAgent: 'code-notes'.concat(packageJson.version),
  mediaType: {
    format: 'application/vnd.github.v3+json',
  },
});
```

Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: leave userAgent, and mediaType there correct? to avoid repetition?

Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: I'm struggling with this suggestion do you want to put a snippet how you want it on jsfiddle or here?

Comment (kodaman2) [reply src/renderer/store/modules/Note.js]: Maybe I got it commit coming thru : )

**Comment by kodaman2** [comment]
Ok, hopefully I understood the last change request.

**Comment by flashadvocate** [comment]
Seems to be working nicely as of the latest commit to this PR

**Comment by kodaman2** [comment]
> Seems to be working nicely as of the latest commit to this PR (tested up to [fa256e4](https://github.com/lauthieb/code-notes/commit/fa256e4d100b3e40c1e443d0f4a352d0442a6f3f))

Cool thanks! it never hurts to check, would hate to merge it and then not work lol. A few more reviews and we should be good for merging. ; )

**Comment by lauthieb** [comment]
Hi @kodaman2, it's almost done :D
Can you please review my Pull Request on your branch here: kodaman2#1

@flashadvocate can you test this Pull Request also please? This is my branch : `ghe-token-refactoring`

**Comment by kodaman2** [comment]
> Hi @kodaman2, it's almost done :D
> Can you please review my Pull Request on your branch here: [kodaman2#1](https://github.com/kodaman2/code-notes/pull/1)
> 
> @flashadvocate can you test this Pull Request also please? This is my branch : `ghe-token-refactoring`

@lauthieb Merged, @flashadvocate you can test on my branche bd8b2daf1225b7cd18950998e5ff940f798586ce

**Comment by kodaman2** [comment]
@lauthieb how exactly do these changes work? I am just wondering, to be honest javascript constructors are super weird to my python/java eyes lol. Will need to read upon ES6 a bit more.

```
const getOctokit = ({ githubPersonalAccessToken, githubEnterpriseUrl }) => new Octokit({
  auth: githubPersonalAccessToken,
  userAgent: 'code-notes'.concat(packageJson.version),
  mediaType: {
    format: 'application/vnd.github.v3+json',
  },
  ...(githubEnterpriseUrl && { baseUrl: githubEnterpriseUrl }),
});

const octokit = getOctokit(store.rootState.Settings.settings);
```

**Comment by lauthieb** [comment]
> @lauthieb how exactly do these changes work? I am just wondering, to be honest javascript constructors are super weird to my python/java eyes lol. Will need to read upon ES6 a bit more.
> 
> ```
> const getOctokit = ({ githubPersonalAccessToken, githubEnterpriseUrl }) => new Octokit({
>   auth: githubPersonalAccessToken,
>   userAgent: 'code-notes'.concat(packageJson.version),
>   mediaType: {
>     format: 'application/vnd.github.v3+json',
>   },
>   ...(githubEnterpriseUrl && { baseUrl: githubEnterpriseUrl }),
> });
> 
> const octokit = getOctokit(store.rootState.Settings.settings);
> ```

First, you have a [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) for settings when I get `githubPersonalAccessToken` and `githubEnterpriseUrl`.
Then we use the `new` keyword as Octokit documentation said.
Finally, I use a [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) and `&&` logical AND, in this way, this line replace the if statement. If `githubEnterpriseUrl` is defined, then we can set a new key in this object named `baseUrl` spreaded in the object.
I found this tip there : https://stackoverflow.com/a/40560953

**Comment by kodaman2** [comment]
Thanks for the explanation. Makes sense, looks pretty lean too!

**Comment by flashadvocate** [comment]
Working great as of bd8b2daf1225b7

**Comment by lauthieb** [comment]
Merged!
I will make a new version as soon as possible :)

**Comment by lauthieb** [comment]
Thanks guys for this work.

**Comment by flashadvocate** [comment]
> Merged!
> I will publish a new version as soon as possible :)

Any chance for a release?

**Comment by lauthieb** [comment]
Hi @flashadvocate, I will do that on Saturday.

**Comment by lauthieb** [comment]
Here is the new release 1.2.3 of Code Notes : https://github.com/lauthieb/code-notes/releases/tag/1.2.3 :tada:
Congrats @flashadvocate & @kodaman2!

**Pull request merged**